fix: Implement mixed site hygiene

ChayimFriedman2 opened the pull request:
The realization that made me do this (as explained in detail in the first commit) is that while def-site hygiene is incredibly hard for us, mixed-site hygiene should be easy (well, I underestimated it due to the need of the second commit, but still). And the best part in that is that mixed-site hygiene rules: def-site hygiene is available only on nightly and without a clear path to stabilization, and even when (if) it will stabilize this will still help solving the case of variables/labels.

This brings diagnostics on self down to four (from five, I think? or more) 🎉 and the remaining diagnostics all require next solver (three due to coercion issue, one due to the issue of trait bounds not being preferred).

Best reviewed commit by commit.

Fixes #18262.
Fixes #16342.
Fixes #13758.
Fixes #11681.
Fixes #10535.

(Wow, so many issues! That's heart-filling :heart:)

Veykril commented on crates/hir-def/src/body.rs:
```diff
@@ -53,6 +69,20 @@ pub struct Body {
     pub body_expr: ExprId,
     /// Block expressions in this body that may contain inner items.
     block_scopes: Vec<BlockId>,
+
+    /// A map from binding to its hygiene ID.
+    ///
+    /// Bindings that don't come from macro expansion are not allocated to save space, so not all bindings appear here.
+    /// If a binding does not appear here it has `SyntaxContextId::ROOT`.
+    ///
+    /// Note that this may not be the direct `SyntaxContextId` of the binding's expansion, because transparent
+    /// expansions are attributed to their parent expansion (recursively).
+    binding_hygiene: FxHashMap<BindingId, HygieneId>,
+    /// A map from an variable usages to their hygiene ID.
+    ///
+    /// Expressions that can be recorded here are single segment path, although not all single segments path refer
+    /// to variables and have hygiene (some refer to items, we don't know at this stage).
+    expr_hygiene: FxHashMap<ExprId, HygieneId>,
```
I guess this wouldnt be necessary if our `Name`s were storing the `SyntaxContextId`? Given path expressions would then have hygienic paths, bindings also already containing `Name`s, is there a reason we are doing it this way instead of giving `Name`'s its `SyntaxContextId` (which currently is just a `()`)?

ChayimFriedman2 replied:
Yes, this won't be necessary, but this is the entire point of this PR - giving every `Name` hygiene, while required for def-site hygiene, is likely to be much more intrusive and introduce performance penalty, so we instead do it only for variables/labels.

Veykril replied:
Ah okay, that was the reasoning behind it gotcha

Veykril replied:
Unresolving this once, can we drop a FIXME *somewhere* talking about this decision just so if someone comes in and fully implements hygiene that they know why this special stuff exists right now?

ChayimFriedman2 replied:
Done.

Veykril commented on crates/hir-def/src/body/scope.rs:
```diff
@@ -22,6 +22,7 @@ pub struct ExprScopes {
 #[derive(Debug, PartialEq, Eq)]
 pub struct ScopeEntry {
     name: Name,
+    hygiene: HygieneId,
```
(just noting things here now) This wouldn't be needed if `Name` has its syntax context populated either. Though I guess here it would be duplicated, once in `name: Name` and once in the `Binding` referred to by the `binding: BindingId`

Veykril commented on crates/hir-def/src/expander.rs:
```diff
@@ -49,6 +49,10 @@ impl Expander {
         }
     }
 
+    pub(crate) fn span_map(&self) -> Option<&SpanMap> {
+        self.span_map.get()
+    }
```
This seems a bit weird of an API, this shouldnt be optional (it only is due to not always being initialized given the lazy OnceCell)

Veykril replied:
We also need to be careful with the span map as that currently accessing it unfortunately completely breaks incrementality :( Though it probably is fine here (as bodies are invalidated anyways), and type inference only when its actually using macros then (assuming we only access it then)

ChayimFriedman2 replied:
I only need this after expanding a macro (to access the new span map), and a span map is always available in this case, so it doesn't really matter if it returns an `Option` or initializes the cell. At first I tried making `enter_expand()` return the span map too, but this is a more intrusive change (even if it saves an unwrap).

Also inference can't access this since this is `pub(crate)`.

Veykril commented on crates/hir-def/src/body/lower.rs:
```diff
@@ -60,6 +62,13 @@ pub(super) fn lower(
+    // even though they should be the same. Also, when the body comes from multiple expansions, their
+    // hygiene is different.
+    let span_map =
+        expander.current_file_id().macro_file().map(|file_id| db.expansion_span_map(file_id));
```
The expander contains a `span_map: OnceCell<SpanMap>`, we can populate that in this case

Veykril commented on crates/hir-expand/src/lib.rs:
```diff
@@ -217,8 +220,27 @@ pub struct MacroCallLoc {
 }
 impl InternValueTrivial for MacroCallLoc {}
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct Macro2Id(pub salsa::InternId);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct MacroRulesId(pub salsa::InternId);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct ProcMacroId(pub salsa::InternId);
+
+/// A macro
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum MacroId {
+    Macro2Id(Macro2Id),
+    MacroRulesId(MacroRulesId),
+    ProcMacroId(ProcMacroId),
+}
+impl_from!(Macro2Id, MacroRulesId, ProcMacroId for MacroId);
+
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub struct MacroDefId {
+    pub id: MacroId,
```
Instead of doing this, can we just resolve the `MacroId` to its `MacroDefId` for now via `DefDatabase::macro_def`?

ChayimFriedman2 replied:
Hmm... `MacroDefId` contains an `AstId`. Won't it hurt incrementality?

Veykril replied:
Only if we store it in some query result again which I don't think is the case? (will double check that tomorrow)

ChayimFriedman2 replied:
That means we will have to store `MacroDefId` in body, and then access it during it inference etc. Although macro definitions in bodies are quite rare, so maybe it's okay.

Veykril replied:
Ah we store this in the `ScopeData` right, I missed that. That still shouldn't be a problem though I think, if the Macro's AstId changes we recompute a bunch of base queries anyways so we aren't losing out on incrementality here. (still looking into this more though to see if there is other problems)

Veykril replied:
https://github.com/Veykril/rust-analyzer/commit/883c4b9c9dd3328bd3678546cd2a9b6d31ffa865 is the change changing this back, from what I can tell this should be fine (aside from `ScopeData` becoming bigger, that might be a bit much)

Main reason I don't like the change from your PR here is that it makes the Macro ID stuff even more confusing (it is already a nightmare to grasp imo), I'd love if we could simplify it somehow but that likely should wait for the next salsa either way.

ChayimFriedman2 replied:
Okay, I changed this to store `MacroDefId`.

bors commented:
:umbrella: The latest upstream changes (presumably #18239) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Veykril commented on crates/hir-def/src/body/lower.rs:
```diff
@@ -92,6 +103,12 @@ struct ExprCollector<'a> {
 
+    /// Legacy (`macro_rules!`) macros can have multiple definitions and shadow each other,
+    /// and we need to find the current definition. So we track the number of definitions we saw.
+    current_block_legacy_macro_defs_count: FxHashMap<Name, usize>,
```
I wonder, does this work properly across blocks (that is shadowing in different nesting levels? Feels like it should but just double checking here. Aside, we still need to figure out something similar for def map name resolution, though macro resolution in there is generally too messy right now

ChayimFriedman2 replied:
Yes, because each block will have a new defmap. I added a test.

Veykril commented on crates/hir-def/src/body/lower.rs:
```diff
@@ -1145,10 +1206,44 @@ impl ExprCollector<'_> {
+                    .scope
+                    .get(&name)
+                    .take_macros()
+                    .expect("def map should have macro definition, but it doesn't");
```
```suggestion
                    .expect("def map should have macro definition, but it doesn't");
```
Might be better to skip with a `never!` instead, this feels like we could easily write a bug that reaches this

Veykril reviewed (commented):
Okay one more last thing, does this have a significant impact on memory (sorry, I'll keep asking this 😬), and does this have a noticable perf impact on the `integrated_benchmark` benches?

ChayimFriedman2 commented:
@Veykril Addressed comments.

About perf impact:

This does not have a significant impact on memory (nor do I expect it to have after the removal of the `MacroId` from `MacroDefId`), around 5mb but that's in the unstable range.

For speed, the integrated benchmarks do not show a meaningful difference, but they are very flaky. If I benchmark the whole thing (including workspace loading) using hyperfine, it is up to 200ms slower, but I tend to think this should be attributed to the workspace loading and not the benchmark itself, if this is even a true regression and not just a measurement error..

`analysis-stats .` is not affected.
